### PR TITLE
Add validation tool

### PR DESCRIPTION
**Pull request by jfkonecn**
resolves oncokb/oncokb#3703

### CODE REVIEW

File: src/main/java/org/mskcc/oncokb/curation/web/rest/ClientForwardController.java [hunk 1]
```diff
@@ -40,7 +40,7 @@ public ResponseEntity<String> index() {
      * Forwards any unmapped paths (except those containing a period) to the client {@code index.html}.
      * @return forward to client {@code swagger-ui/index.html}.
      */
-    @GetMapping(value = "/**/{path:[^\\.]*}")
+    // @GetMapping(value = "/**/{path:[^\\.]*}")
     public String forward() {
```
Comment (jfkonecn) [inline src/main/java/org/mskcc/oncokb/curation/web/rest/ClientForwardController.java:44]: Need to fix this. It's blocking websocket from working.

**Comment by zhx828** [comment]
Some feature requests before code review
- for success test/info, we should not allow the collapsible to open. See deleted `test` mutation in https://curation.oncokb.org/curation/ZNRF3/somatic as an example
- we need to notify the user the validation is done. Maybe add a toast banner?
- Let's change the Variant column header to `Biomarker`
- the Validate button does not need to be wrapped in a container again. Otherwise you will see extra white space on the left of the button comparing to Tools tab
- I don't think the refresh icon should be shown up before the first time of validation. And the button should be primary button before first time of validation. See `Release Gene` button in the gene page under the Tools tab

File: src/main/webapp/app/components/tabs/CurationDataValidationTab.tsx [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    error: true,
+  });
+  useEffect(() => {
+    onChange(validationResultFilter);
+  }, [validationResultFilter]);
+
+  return (
+    <Form className={className}>
+      <Row>
+        <Col>
```
Comment (bprize15) [inline src/main/webapp/app/components/tabs/CurationDataValidationTab.tsx:34]: Could you remove the padding on the column so it lines up nicely with the tab titles?

File: src/main/webapp/app/components/tabs/CurationDataValidationTab.tsx [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+  const [isValidating, setIsValidating] = React.useState(false);
+  const [validationCounter, setValidationCounter] = React.useState(0);
+
+  useEffect(() => {
+    setIsValidating(validationCounter > 0);
+  }, [validationCounter]);
+
+  const onClose = useCallback(() => {
+    setIsValidating(false);
+    toast.success('Validation completed!');
```
Comment (bprize15) [inline src/main/webapp/app/components/tabs/CurationDataValidationTab.tsx:239]: We have utility functions `notifySuccess` and `notifyError` for this

File: src/main/webapp/app/components/tabs/CurationDataValidationTab.tsx [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+  }, []);
+  const onOpen = useCallback(() => {
+    return;
+  }, []);
+  const onError = useCallback(() => {
+    toast.error('Error while attempting to validate. Please try again.');
```
Comment (bprize15) [inline src/main/webapp/app/components/tabs/CurationDataValidationTab.tsx:245]: Can use `notifyError` here

File: src/main/webapp/app/pages/curation/collapsible/Collapsible.tsx [hunk 5]
```diff
@@ -25,17 +34,41 @@ export default function Collapsible({
...
   const displayOptionsOverride = useMemo(() => {
+    memoRerenderRef.current++;
     if (isPendingDelete) {
       defaultDisplayOptions.disableCollapsible = true;
       defaultDisplayOptions.hideAction = displayOptions?.hideAction || false;
       defaultDisplayOptions.hideInfo = displayOptions?.hideInfo || false;
     }
+    const displayOptionsJson = JSON.stringify(displayOptions);
+    const oldDisplayOptionsJson = JSON.stringify(displayOptionRef.current);
+    if (isPendingDeleteRef.current === isPendingDelete && displayOptionsJson === oldDisplayOptionsJson) {
```
Comment (bprize15) [inline src/main/webapp/app/pages/curation/collapsible/Collapsible.tsx:52]: Can delete this I'd imagine

File: .nvmrc [hunk 6]
```diff
@@ -0,0 +1 @@
+14.17.6
```
Comment (zhx828) [inline .nvmrc:1]: I noticed this is added in few of your prs. When .nvmrc file is specified, nvm will switch to the version automatically?

File: src/main/docker/keycloak.yml [hunk 7]
```diff
@@ -1,30 +1,40 @@
...
+      POSTGRES_DB: keycloak
+      POSTGRES_USER: root
+      POSTGRES_PASSWORD: root
     volumes:
-      - ./realm-config:/opt/jboss/keycloak/realm-config
-      # Keycloak allows you to create custom themes: https://www.keycloak.org/docs/latest/server_development/#_themes
-      # - ./realm-config/themes/custom-theme:/opt/jboss/keycloak/themes/custom-theme
+      - kc-db:/var/lib/mysql
+    networks:
+      - kc-network
```
Comment (zhx828) [inline src/main/docker/keycloak.yml:14]: Curious why we create a new network

File: src/main/java/org/mskcc/oncokb/curation/web/websocket/ProxyCurationValidationApiHandler.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+public class ProxyCurationValidationApiHandler implements WebSocketHandler {
+
+    private OncoCoreWebSocketHandler oncoHandler;
+    private String validationEndpoint;
+
+    public ProxyCurationValidationApiHandler(ApplicationProperties applicationProperties) {
+        String url = applicationProperties.getOncokb().getUrl();
+        url = url.replace("https://", "wss://");
+        url = url.replace("http://", "ws://");
+        this.validationEndpoint = url + "/api/websocket/curation/validation";
```
Comment (zhx828) [inline src/main/java/org/mskcc/oncokb/curation/web/websocket/ProxyCurationValidationApiHandler.java:22]: Should we move this to the application as well?

File: src/main/webapp/app/components/tabs/CurationDataValidationTab.tsx [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+    setIsValidating(false);
+    notifySuccess('Validation completed!');
+  }, []);
+  const onOpen = useCallback(() => {
+    return;
+  }, []);
+  const onError = useCallback(() => {
+    notifyError(new Error('Error while attempting to validate. Please try again.'));
+  }, []);
+  const url = relativeWebsocketUrlToAbsolute('/api/websocket/curation/validation');
```
Comment (zhx828) [inline src/main/webapp/app/components/tabs/CurationDataValidationTab.tsx:247]: I would prefer creating a websocket API in curation platform and let the server side routes certain URL pattern to the core service so we could also add other websockets as well.

**Review by zhx828 — CHANGES_REQUESTED** [review]
For some reason, the validation websocket doesn't work after recent changes. I will test again.

**Comment by zhx828** [comment]
Also thanks for testing useMemo. I will take another look at it when possible. What's the conclusion on that?

**Comment by jfkonecn** [comment]
> Also thanks for testing useMemo. I will take another look at it when possible. What's the conclusion on that?

It rerenders every time.
According to the [react docs](https://react.dev/reference/react/useMemo) it uses [Object.is()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) to do the equality check which does a pointer check for the equality between two objects.

**Review by zhx828 — APPROVED** [review]
Could you create a ticket to follow through on swagger/util as we discussed? Thanks!

**Pull request merged**